corehq/blobs/interface.py

```diff
@@ -61,6 +62,25 @@ def get(self, key):
         """
         raise NotImplementedError
 
+    def _validate_get_args(self, key, type_code, meta):
+        if type_code == CODES.form_xml:
+            if meta is None:
+                raise ValueError("form XML must be loaded with 'meta' argument")
+            elif key is not None:
+                raise ValueError("'key' and 'meta' are mutually exclusive")
+            elif type_code is not None:
+                raise ValueError("'type_code' and 'meta' are mutually exclusive")
+            return meta.key
+        elif key is not None or type_code is not None:
+            if key is None or type_code is None:
+                raise ValueError("'key' must be specified with 'type_code'")
+            if meta is not None:
+                raise ValueError("'key' and 'meta' are mutually exclusive")
+            return key
+        if meta is None:
+            raise ValueError("'key' and 'type_code' or 'meta' is required")
+        return meta.key
```

Comment:
I think the version here is more concise: https://github.com/dimagi/commcare-hq/issues/26813#issuecomment-600800727

Reply:
Following issues with that version:
- If function is called with `type_code == "form_xml"` and one of `key` or `error`, they will get following error:
```
'key' must be specified with 'type_code'
```
whereas what they really should get is
```
form XML must be loaded with 'meta' argument
```
Current version, although longer, handles all possible cases with the correct error message

Reply:
I didn't follow this. Are you saying this function call
```py
validate_get_args("some-key", CODES.form_xml, None)
```
will raise `'key' must be specified with 'type_code'`?

That's not what I see. Tested locally this is what I get:
```
Traceback (most recent call last):
  File "<string>", in validate_get_args
ValueError: form XML must be loaded with 'meta' argument

```

Reply:
No, I am referring to the version at https://github.com/dimagi/commcare-hq/issues/26813#issuecomment-600800727 since @snopoke suggested using it.

Using that version will raise `'key' must be specified with 'type_code'` with following function call:

validate_get_args("some-key", CODES.form_xml, None)

Reply:
Might want to check that again. I ran the version from https://github.com/dimagi/commcare-hq/issues/26813#issuecomment-600800727 and got `ValueError: form XML must be loaded with 'meta' argument`

```py
>>> class CODES:
...     # stub
...     form_xml = 2
...
>>> def validate_get_args(key, type_code, meta):
...     if key is not None or type_code is not None:
...         if key is None or type_code is None:
...             raise ValueError("'key' must be specified with 'type_code'")
...         if meta is not None:
...             raise ValueError("'key' and 'meta' are mutually exclusive")
...         if type_code == CODES.form_xml:
...             raise ValueError("form XML must be loaded with 'meta' argument")
...         return key
...     if meta is None:
...         raise ValueError("'key' and 'type_code' or 'meta' is required")
...     return meta.key
...
>>> validate_get_args("some-key", CODES.form_xml, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 8, in validate_get_args
ValueError: form XML must be loaded with 'meta' argument
```

Reply:
Ok, I checked again. But consider below

>>> class CODES:
...     form_xml = 2
... 
>>> def validate_get_args(key, type_code, meta):
...     if key is not None or type_code is not None:
...         if key is None or type_code is None:
...             raise ValueError("'key' must be specified with 'type_code'")
...         if meta is not None:
...             raise ValueError("'key' and 'meta' are mutually exclusive")
...         if type_code == CODES.form_xml:
...             raise ValueError("form XML must be loaded with 'meta' argument")
...         return key
...     if meta is None:
...         raise ValueError("'key' and 'type_code' or 'meta' is required")
...     return meta.key
... 
>>> validate_get_args(None, CODES.form_xml, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in validate_get_args
ValueError: 'key' must be specified with 'type_code'

Reply:
Ok, I checked again. But consider below

```
>>> class CODES:
...     form_xml = 2
... 
>>> def validate_get_args(key, type_code, meta):
...     if key is not None or type_code is not None:
...         if key is None or type_code is None:
...             raise ValueError("'key' must be specified with 'type_code'")
...         if meta is not None:
...             raise ValueError("'key' and 'meta' are mutually exclusive")
...         if type_code == CODES.form_xml:
...             raise ValueError("form XML must be loaded with 'meta' argument")
...         return key
...     if meta is None:
...         raise ValueError("'key' and 'type_code' or 'meta' is required")
...     return meta.key
... 
>>> validate_get_args(None, CODES.form_xml, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in validate_get_args
ValueError: 'key' must be specified with 'type_code'
```

Reply:
Looks right to me (you sent a null key with a type code). Do you see a problem?

Reply:
"We sent `type_code` as `form_xml`. `key` is not expected for this type_code and only `meta` should be provided. Yet the error asks me to provide `key`."

Reply:
This should fix it
```py
def validate_get_args(key, type_code, meta):
    if key is not None or type_code is not None:
        if meta is not None:
            raise ValueError("'key' and 'meta' are mutually exclusive")
        if type_code == CODES.form_xml:
            raise ValueError("form XML must be loaded with 'meta' argument")
        if key is None or type_code is None:
            raise ValueError("'key' must be specified with 'type_code'")
        return key
    if meta is None:
        raise ValueError("'key' and 'type_code' or 'meta' is required")
    return meta.key
```